shared/src/main/scala/coop/rchain/sdk/dag/data/DagData.scala

```diff
@@ -0,0 +1,21 @@
+package coop.rchain.sdk.dag.data
+
+trait DagData[M, S] {
```

Comment:
I was trying to implement things like this several months before and I think `trait` like this is kind of over-abstract from my perspective. The abstraction gives very high flexibility indeed while the implementation of the trait seems more codes and long.

I prefer something like 
```scala
trait DagData {
  def seqNum: Long
  def blockNum: Long
  def justifications: Set[M]
  def sender[S]: S # or put S in DagData[S], not sure which is better
  def bondsMap: Map[S, Long]
  def stake[S](s: S): Long
}
```
and when we need type works like `DagData`, we could use like below
```
trait DagManager[F[_], M <: DagData, S] {
  def getDagView(latestMessages: Set[M]): F[DagView[F, M, S]]
  def updateDag(m: M): F[Unit]
}
```

In this way, we can build things with some relation to each other but these constraints might be blocked in some way.

Reply:
Actually, after going over all the codes, I think this idea can hold back and try with what you have.